opentracing-api/src/main/java/io/opentracing/Tracer.java

```diff
@@ -20,73 +20,80 @@
 public interface Tracer {
 
   /**
-   * Create, start, and return a new Span with the given `operationName`.
-   * An optional parent Span can be specified used to incorporate the newly-returned Span into an existing trace.
+   * Return a new SpanBuilder for a Span with the given `operationName`.
    *
-   * <p>Example:
+   * <p>A contrived example:
    * <pre>{@code
     Tracer tracer = ...
 
-    Span feed = tracer.buildSpan("GetFeed")
-                      .start();
+    Span parentSpan = tracer.buildSpan("DoWork")
+                            .start();
 
     Span http = tracer.buildSpan("HandleHTTPRequest")
-                      .withParent(feed)
+                      .withChildOf(parentSpan.context())
                       .withTag("user_agent", req.UserAgent)
                       .withTag("lucky_number", 42)
                       .start();
     }</pre>
    */
   SpanBuilder buildSpan(String operationName);
 
-  /** Takes two arguments:
-   *    a Span instance, and
-   *    a “carrier” object in which to inject that Span for cross-process propagation.
+  /**
+   * Inject a SpanContext into a `carrier` of a given `format`, presumably for propagation across process boundaries.
+   *
+   * <p>Example:
+   * <pre>{@code
+   * Tracer tracer = ...
+   * Span clientSpan = ...
+   * TextMapWriter httpHeaderWriter = new AnHttpHeaderCarrier(httpRequest);
+   * tracer.inject(span.context(), BuiltinFormats.HTTP_HEADER, httpHeaderWriter);
+   * }</pre>
    *
-   * A “carrier” object is some sort of http or rpc envelope, for example HeaderGroup (from Apache HttpComponents).
+   * @param spanContext the SpanContext instance to inject into the carrier
+   * @param format the Format of the carrier. See BuiltinFormats.
+   * @param carrier the carrier for the SpanContext state; when inject() returns, the Tracer implementation will have represented the SpanContext within `carrier`
    *
-   * Attempting to inject to a carrier that has been registered/configured to this Tracer will result in a
-   * IllegalStateException.
+   * All implementations must support, at minimum, the BuiltinFormats.
    *
-   * All implementations support at minimum the required carriers BinaryWriter and TextMapWriter.
+   * @see BuiltinFormats
    */
-  <T> void inject(Span span, T carrier);
+  <W> void inject(SpanContext spanContext, Format<?, W> format, W carrier);
```

Comment:
The Format thing, which accepts any input and uses `?` as one of the parameters is feigning type safety. There's nothing that binds R and W together when `?` is used, right?

I'm most concerned about how this api is designed, not what it is trying to solve.

I think if you posted to a larger audience, they would agree that this api is tedious, almost scala in java:

```java
<W> void inject(SpanContext spanContext, Format<?, W> format, W carrier);
```

If you need to express a qualifier, you can append it as a parameter, but I wouldn't use a dual-parameter qualifier key knowing one would be ignored. If the only use-case is http, and the only case is baggage, I'd at least consider finding another way.

Most importantly, when designing a new api we should cite other similar designs, especially when they are full of types.

For example, there's a similar concern in dependency injection. They `Key` type below is made to both hold a qualifier and also the type of the instance passed in, (in this case `W`). It has more responsibility than what we are trying to do, yet has a less complex signature. Let's try to simplify!

https://google.github.io/guice/api-docs/latest/javadoc/index.html?com/google/inject/Key.html